RFCs/FS-1152-For-loop-with-accumulation.md

```diff
@@ -0,0 +1,272 @@
+- More precise error messages from type inference especially for newcomers
+
+The only caveat is the unfamiliarity with a loop that has a return value where existing loops must return unit, but over time, this can be overcome with stressing the existence of an accumulator vs not having one.
+
+# Detailed design
+
+For this syntax
+
+```fs
+for <pat1> = <expr1> with <pat2> in <expr2> do
+    <exprs>
+    <expr3>
+```
+
+it undergoes a simple syntactical expansion to
+
+```fs
+let mutable <accum> = <expr1>
+for <pat2> in <expr2> do
+    let <pat1> = <accum>
+    <exprs>
+    <accum> <- <expr3>
+<accum>
+```
+where `<accum>` is a compiler-generated accumulation variable that cannot be used outside the loop,
+and `<exprs>` match any expressions inside a sequence expression before the final expression in the loop body,
+or any `if` and `match` and `try`-`with` permitting computation expression syntax inside.
+
+If the final expression `<expr3>` is contained inside `if` or `match` or `try`-`with`, then each branch must return the same value and `<expr3>` is searched recursively with the final expression of each branch for `<accum> <-` application.
+
+The result of the syntactical translation shows that `<accum>` is the value of this loop. It is also the accumulator which gets updated each iteration.
+
+Note that `<exprs>` are subject to usual treatment by computation expressions including application of implicit yields,
+which is different from the final expression `<expr3>`.
+
+Therefore, this may happen:
+```fs
+let result = [
+    for sentence = "" with word in ["Hello"; " "; "World"; "!"] do
+        sentence
+        sentence + word
+    |> printfn "%s"
+]
+// expands to
+let result = [
+    let mutable <accum> = ""
+    for word in ["Hello"; " "; "World"; "!"] do
+        sentence // Implicit yield here
+        <accum> <- sentence + word // But not for the final expression of the loop body
+    <accum>
+    |> printfn "%s"
+]
+```
+
+### Unresolved question
+
+There are a few choices that can be made for `<exprs>` -
+- Allow implicit yields before the final value
+- Warn with implicit yield behaviour
+- Warn with discard behaviour
+- Error because it's ambiguous
+
+Allowing implicit yields before the final value might not be the best choice here.
+
+Warning with discard is also a valid choice when inside this kind of for loop. Explicit `yield`s are always available if needed inside this loop, and if we choose to warn, explicit `yield`s in this kind of loop shouldn't interfere with the ability to use implicit `yield`s outside.
+
+# Drawbacks
+
+It's another syntax to learn.
+
+# Alternatives
+
+Doing nothing - we miss an opportunity to bridge regular for loops with lambda folds.
+
+There is a potential extension to folding across multiple sequences with the `and` keyword in the future -
+```fs
+for s, p = 0, 0 with t in ts and u in us do
+    s + t, p + u
+```
```

Comment:
This example gives the impression that folding over N sequences would require N accumulators, but that's not the case, for example:

```fsharp
// n-dimensional Euclidian distance between A and B

let pointACoordinates = [ 1.5; 2.; -3.4; -1.2 ]
let pointBCoordinates = [ -3.1; 3.1; 1.; -0.2 ]

let euclidianDistance =
    for acc = 0. with an in pointACoordinates and bn in pointBCoordinates do
        acc + (an - bn) ** 2.
    |> sqrt
```